In [1]:
# import all packages and set plots to be embedded inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [2]:
# For better assessment set pandasDataFrame not to collapse rows and columns when displaying them
pd.set_option('display.max_rows', None, 'display.max_columns', None)

In [3]:
# Error:
# SettingWithCopyWarning: A value is trying to be set on a copy of a
# slice from a DataFrame

# As explained in the Source, this warning is usually safe to ignore. You
# can disable it by running the following:
pd.options.mode.chained_assignment = None  # default='warn'

In [4]:
# load in the dataset into a pandas dataframe and make a copy of the dataFrame
loan_df = pd.read_csv('prosperLoanData.csv')
prosper = loan_df.copy()

##  High Level Assessment of the Data Shape and Composition

In [25]:
prosper.head(10)

creditgrade  term loanstatus  borrowerapr  borrowerrate  lenderyield  \
0           C    36  Completed      0.16516        0.1580       0.1380   
1          NA    36    Current      0.12016        0.0920       0.0820   
2          HR    36  Completed      0.28269        0.2750       0.2400   
3          NA    36    Current      0.12528        0.0974       0.0874   
4          NA    36    Current      0.24614        0.2085       0.1985   
5          NA    60    Current      0.15425        0.1314       0.1214   
6          NA    36    Current      0.31032        0.2712       0.2612   
7          NA    36    Current      0.23939        0.2019       0.1919   
8          NA    36    Current      0.07620        0.0629       0.0529   
9          NA    36    Current      0.07620        0.0629       0.0529   

   estimatedeffectiveyield  estimatedloss  estimatedreturn  \
0                 0.168661            NaN              NaN   
1                 0.079600         0.0249          0.05470   
2                 0.168661            NaN              NaN   
3                 0.084900         0.0249          0.06000   
4                 0.183160         0.0925          0.09066   
5                 0.115670         0.0449          0.07077   
6                 0.238200         0.1275          0.11070   
7                 0.178300         0.0799          0.09840   
8                 0.052210         0.0099          0.04231   
9                 0.052210         0.0099          0.04231   

   prosperrating (numeric) prosperrating (alpha)  prosperscore  \
0                      NaN                   NaN           NaN   
1                      6.0                     A           7.0   
2                      NaN                   NaN           NaN   
3                      6.0                     A           9.0   
4                      3.0                     D           4.0   
5                      5.0                     B          10.0   
6                      2.0                     E           2.0   
7                      4.0                     C           4.0   
8                      7.0                    AA           9.0   
9                      7.0                    AA          11.0   

   listingcategory (numeric) borrowerstate employmentstatus  \
0                          0            CO    Self-employed   
1                          2            CO         Employed   
2                          0            GA    Not available   
3                         16            GA         Employed   
4                          2            MN         Employed   
5                          1            NM         Employed   
6                          1            KS         Employed   
7                          2            CA         Employed   
8                          7            IL         Employed   
9                          7            IL         Employed   

   isborrowerhomeowner                 groupkey  creditscorerangelower  \
0                 True                      NaN                  640.0   
1                False                      NaN                  680.0   
2                False  783C3371218786870A73D20                  480.0   
3                 True                      NaN                  800.0   
4                 True                      NaN                  680.0   
5                 True                      NaN                  740.0   
6                False                      NaN                  680.0   
7                False                      NaN                  700.0   
8                 True                      NaN                  820.0   
9                 True                      NaN                  820.0   

   creditscorerangeupper  currentcreditlines  opencreditlines  \
0                  659.0                 5.0              4.0   
1                  699.0                14.0             14.0   
2                  499.0                 NaN              NaN   
3       

In [5]:
prosper.tail(10)

ListingKey  ListingNumber            ListingCreationDate  \
113927  E3433419834735803891976         322787  2008-04-30 21:25:19.670000000   
113928  E34935176664905343E01EA         510097  2011-06-06 19:02:44.443000000   
113929  E3553583161337791FCB87F         830413  2013-07-06 17:40:01.657000000   
113930  E35D3584034795373BCD69A         831523  2013-07-08 10:24:49.700000000   
113931  E36F36005339663245C20F8        1141292  2014-01-16 20:13:08.040000000   
113932  E6D9357655724827169606C         753087  2013-04-14 05:55:02.663000000   
113933  E6DB353036033497292EE43         537216  2011-11-03 20:42:55.333000000   
113934  E6E13596170052029692BB1        1069178  2013-12-13 05:49:12.703000000   
113935  E6EB3531504622671970D9E         539056  2011-11-14 13:18:26.597000000   
113936  E6ED3600409833199F711B7        1140093  2014-01-15 09:27:37.657000000   

       CreditGrade  Term              LoanStatus           ClosedDate  \
113927          AA    36               Completed  2011-05-09 00:00:00   
113928         NaN    36               Completed  2011-09-19 00:00:00   
113929         NaN    36               Completed  2014-02-07 00:00:00   
113930         NaN    36                 Current                  NaN   
113931         NaN    60                 Current                  NaN   
113932         NaN    36                 Current                  NaN   
113933         NaN    36  FinalPaymentInProgress                  NaN   
113934         NaN    60                 Current                  NaN   
113935         NaN    60               Completed  2013-08-13 00:00:00   
113936         NaN    36                 Current                  NaN   

        BorrowerAPR  BorrowerRate  LenderYield  EstimatedEffectiveYield  \
113927      0.07469        0.0679       0.0579                      NaN   
113928      0.22362        0.1899       0.1799                  0.17570   
113929      0.30285        0.2639       0.2539                  0.23180   
113930      0.20053        0.1639       0.1539                  0.14520   
113931      0.15016        0.1274       0.1174                  0.11210   
113932      0.22354        0.1864       0.1764                  0.16490   
113933      0.13220        0.1110       0.1010                  0.10070   
113934      0.23984        0.2150       0.2050                  0.18828   
113935      0.28408        0.2605       0.2505                  0.24450   
113936      0.13189        0.1039       0.0939                  0.09071   

        EstimatedLoss  EstimatedReturn  ProsperRating (numeric)  \
113927            NaN              NaN                      NaN   
113928         0.0610          0.11470                      4.0   
113929         0.1225          0.10930                      2.0   
113930         0.0574          0.08780                      5.0   
113931         0.0424          0.06970                      5.0   
113932         0.0699          0.09500                      4.0   
113933         0.0200          0.08070                      6.0   
113934         0.1025          0.08578                      3.0   
113935         0.0850          0.15950                      4.0   
113936         0.0299          0.06081                      6.0   

       ProsperRating (Alpha)  ProsperScore  ListingCategory (numeric)  \
113927                   NaN           NaN                          4   
113928                     C           6.0                          3   
113929                     E           3.0                          2   
113930                     B           6.0                          1   
113931                     B           6.0                          3   
113932                     C           5.0                          1   
113933                     A           8.0                          7   
113934                     D           3.0                          1   
113935                     C           5.0                          2   
113936                     A   

In [6]:
prosper.shape

(113937, 81)

In [7]:
prosper.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

In [8]:
prosper.describe()

ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.004900        -0.182700   
25%         0.124200                 0.115670       0.042400         0.074080   
50%         0.173000                 0.161500       0.072400         0.091700   
75%         0.240000                 0.224300       0.112000         0.116600   
max         0.492500                 0.319900       0.366000         0.283700   

       ProsperRating (numeric)  ProsperScore  ListingCategory (numeric)  \
count             84853.000000  84853.000000              113937.000000   
mean                  4.072243      5.950067                   2.774209   
std                   1.673227      2.376501                   3.996797   
min                   1.000000      1.000000                   0.000000   
25%                   3.000000      4.000000                   1.000000   
50%                   4.000000      6.000000                   1.000000   
75%                   5.000000      8.000000                   3.000000   
max                   7.000000     11.000000                  20.000000   

       EmploymentStatusDuration  CreditScoreRangeLower  CreditScoreRangeUpper  \
count             106312.000000          113346.000000          113346.000000   
mean                  96.071582             685.567731             704.567731   
std                   94.480605              66.458275              66.458275   
min                    0.000000               0.000000              19.000000   
25%                   26.000000             660.000000             679.000000   
50%                   67.000000             680.000000             699.000000   
75%                  137.000000             720.000000             739.000000   
max                  755.000000             880.000000             899.000000   

       CurrentCreditLines  OpenCreditLines  TotalCreditLinespast7years  \
count       106333.000000    106333.000000               113240.000000   
mean            10.317192         9.260164                   26.754539   
std              5.457866         5.022644                   13.637871   
min              0.000000         0.000000                    2.000000   
25%              7.000000         6.000000                   17.000000   
50%             10.000000         9.000000                   25.000000   
75%             13.000000        12.000000                   35.000000   
max             59.000000        54.000000                  136.000000   

       OpenRevolvingAccounts  OpenRevolvingMonthlyPayment  \
count           113937.00000                113937.000000   
mean                 6.96979                   398.292161   
std                  4.63097                   447.159711   
min                  0.00000                     0.000000   
25%                  4.00000                   114.000000   
50%                  6.00000                   271.000000   
75%                  9.00000                   525.000000   
max                 51.00000                 14985.000000   

       InquiriesLast6Months  Tot

In [9]:
prosper.isnull().sum()

ListingKey                                  0
ListingNumber                               0
ListingCreationDate                         0
CreditGrade                             84984
Term                                        0
LoanStatus                                  0
ClosedDate                              58848
BorrowerAPR                                25
BorrowerRate                                0
LenderYield                                 0
EstimatedEffectiveYield                 29084
EstimatedLoss                           29084
EstimatedReturn                         29084
ProsperRating (numeric)                 29084
ProsperRating (Alpha)                   29084
ProsperScore                            29084
ListingCategory (numeric)                   0
BorrowerState                            5515
Occupation                               3588
EmploymentStatus                         2255
EmploymentStatusDuration                 7625
IsBorrowerHomeowner               

##### This data set contains 113,937 loans with 81 variables on each loan, including loan amount, borrower rate (or interest rate), current loan status, borrower income, and many others. 42 columns contain some null or missing values

#### Some Cleaning

In [10]:
# Change column names to lower case.
prosper.rename(columns = str.lower, inplace = True)

#### Note that we can plot some histogram with the credit grade column to depict the characteristics of the prosper loan customers with regards to loan repayment habits

In [11]:
prosper.creditgrade.value_counts()

C     5649
D     5153
B     4389
AA    3509
HR    3508
A     3315
E     3289
NC     141
Name: creditgrade, dtype: int64

In [12]:
#credit grade has 84984 missing values. This column is represents a variable of ordinal data type and is a record of the credit
#worthiness of the customer. The various levels range from AA which is the highest order through  A,B,C,D,E,HR,AND NC which is 
#the lowest level. 
#We are going to add another level NA which represents not available and use this to fill up the empty rows.
prosper.creditgrade.fillna('NA', inplace = True)

In [13]:
prosper.creditgrade.value_counts()

NA    84984
C      5649
D      5153
B      4389
AA     3509
HR     3508
A      3315
E      3289
NC      141
Name: creditgrade, dtype: int64

In [14]:
columns_to_drop = ['listingkey', 'listingnumber', 'listingcreationdate', 'closeddate', 'occupation', 
                   'employmentstatusduration', 'currentlyingroup', 'datecreditpulled', 'firstrecordedcreditline', 
                   'totalcreditlinespast7years',  'openrevolvingaccounts', 'openrevolvingmonthlypayment', 
                   'inquirieslast6months', 'totalinquiries', 'incomeverifiable', 'incomerange', 'ontimeprosperpayments', 
                   'totalprosperpaymentsbilled', 'prosperpaymentslessthanonemonthlate', 'prosperpaymentsonemonthpluslate', 'investmentfromfriendsamount', 
                   'investors', 'recommendations', 'percentfunded', 'loanoriginationquarter',
                   'loanoriginationdate', 'loanoriginalamount', 'scorexchangeattimeoflisting',]

In [15]:
def drop_columns(dataframe, columns_to_drop):
    for columns in columns_to_drop:
        try:
            i = columns_to_drop.index(columns)
            #print(i)
            dataframe.drop(columns = columns, inplace = True)
        except Exception as e:
            print(i, e) 
    return 


In [16]:
drop_columns(prosper, columns_to_drop)

In [17]:
prosper.shape

(113937, 53)

In [18]:
prosper.head(10)

creditgrade  term loanstatus  borrowerapr  borrowerrate  lenderyield  \
0           C    36  Completed      0.16516        0.1580       0.1380   
1          NA    36    Current      0.12016        0.0920       0.0820   
2          HR    36  Completed      0.28269        0.2750       0.2400   
3          NA    36    Current      0.12528        0.0974       0.0874   
4          NA    36    Current      0.24614        0.2085       0.1985   
5          NA    60    Current      0.15425        0.1314       0.1214   
6          NA    36    Current      0.31032        0.2712       0.2612   
7          NA    36    Current      0.23939        0.2019       0.1919   
8          NA    36    Current      0.07620        0.0629       0.0529   
9          NA    36    Current      0.07620        0.0629       0.0529   

   estimatedeffectiveyield  estimatedloss  estimatedreturn  \
0                      NaN            NaN              NaN   
1                  0.07960         0.0249          0.05470   
2                      NaN            NaN              NaN   
3                  0.08490         0.0249          0.06000   
4                  0.18316         0.0925          0.09066   
5                  0.11567         0.0449          0.07077   
6                  0.23820         0.1275          0.11070   
7                  0.17830         0.0799          0.09840   
8                  0.05221         0.0099          0.04231   
9                  0.05221         0.0099          0.04231   

   prosperrating (numeric) prosperrating (alpha)  prosperscore  \
0                      NaN                   NaN           NaN   
1                      6.0                     A           7.0   
2                      NaN                   NaN           NaN   
3                      6.0                     A           9.0   
4                      3.0                     D           4.0   
5                      5.0                     B          10.0   
6                      2.0                     E           2.0   
7                      4.0                     C           4.0   
8                      7.0                    AA           9.0   
9                      7.0                    AA          11.0   

   listingcategory (numeric) borrowerstate employmentstatus  \
0                          0            CO    Self-employed   
1                          2            CO         Employed   
2                          0            GA    Not available   
3                         16            GA         Employed   
4                          2            MN         Employed   
5                          1            NM         Employed   
6                          1            KS         Employed   
7                          2            CA         Employed   
8                          7            IL         Employed   
9                          7            IL         Employed   

   isborrowerhomeowner                 groupkey  creditscorerangelower  \
0                 True                      NaN                  640.0   
1                False                      NaN                  680.0   
2                False  783C3371218786870A73D20                  480.0   
3                 True                      NaN                  800.0   
4                 True                      NaN                  680.0   
5                 True                      NaN                  740.0   
6                False                      NaN                  680.0   
7                False                      NaN                  700.0   
8                 True                      NaN                  820.0   
9                 True                      NaN                  820.0   

   creditscorerangeupper  currentcreditlines  opencreditlines  \
0                  659.0                 5.0              4.0   
1                  699.0                14.0             14.0   
2                  499.0                 NaN              NaN   
3       

In [19]:
mean = prosper['estimatedeffectiveyield'].mean()
prosper['estimatedeffectiveyield'].fillna(mean, inplace = True)

In [20]:
prosper.creditgrade.isnull().sum()

0

In [21]:
prosper.loanstatus.isnull().sum()

0

points to note
> credit grade, prosper rating, prosper score

In [22]:
prosper.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 53 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   creditgrade                         113937 non-null  object 
 1   term                                113937 non-null  int64  
 2   loanstatus                          113937 non-null  object 
 3   borrowerapr                         113912 non-null  float64
 4   borrowerrate                        113937 non-null  float64
 5   lenderyield                         113937 non-null  float64
 6   estimatedeffectiveyield             113937 non-null  float64
 7   estimatedloss                       84853 non-null   float64
 8   estimatedreturn                     84853 non-null   float64
 9   prosperrating (numeric)             84853 non-null   float64
 10  prosperrating (alpha)               84853 non-null   object 
 11  prosperscore              

In [23]:
empty = prosper['prosperrating (alpha)'].isnull()

In [24]:
prosper['type'] = prosper['creditgrade'][prosper['prosperrating (alpha)'][empty]]

KeyError: "None of [Float64Index([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,\n              ...\n              nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],\n             dtype='float64', length=29084)] are in the [index]"

In [ ]:
prosper['new_col'].fillna(prosper['prosperrating (alpha)'], inplace = True)

In [ ]:
prosper['new_col'].fillna(prosper['creditgrade'], inplace = True)

In [ ]:
prosper['new_col'] = prosper['prosperrating (alpha)'][prosper['creditgrade']== NaN]

In [ ]:
prosper['new_col'].isnull().sum()